docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+
+> If the peer is `Inactive`, it is set as active.
+> This action has no effect apart from producing metrics.
+
+If a connection slot was upgraded to give room for the established connection, the
+peer on that slot transitions to the [evict sub-state](#evict-peer).
+
+#### Errors
+
+The transition fails if:
+
+- the node dialed itself
+- the peer is already in the `Connected` state
+- the node is connected to `MaxConnected` peers, and no slot is suitable for upgrading
+
+Errors are also returned if:
+
+- the dialed peer was removed from the peer store
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### DialFailed transition
+
+This transition informs a failure when establishing an `outgoing` connection to
+a peer.
+
+The dialed address's `LastDialFailure` time is set, and its `DialFailures`
+counter is increased.
+This information is used to compute the [retry delay](#retry-delay) for the
+dialed address.
+
+The peer manager then spawns a routine that after the computed retry delay
+notifies the next peer to dial routine about the availability of this peer.
+Until then, the peer is the `Frozen Candidate` state.
+
+#### Retry delay
+
+The retry dial is the minimum time, from the latest failed dialing attempt, we
+should wait until dialing a peer address again.
+
+The default delay is defined by `MinRetryTime` parameter.
+If it is set to zero, we *never* retry dialing a peer address.
+
+Upon each failed dial attempt, we increase the delay by `MinRetryTime`, plus an
+optional random jitter of up to `RetryTimeJitter`.
+
+The retry delay should not be longer than the `MaxRetryTime` parameter,
+or `MaxRetryTimePersistent` parameter in the case of persistent peers.
+
+#### Errors
+
+Errors are also returned if:
+
+- the updated peer information is invalid
+- there is an error when saving the peer state to the peer store
+
+### Accepted transition
+
+This transition is performed when the node establishes an *incoming* connection
+with a peer.
+This means that the router has received a connection attempt from this peer and
+successfully established a connection with it.
+
+It may occur, however, that when this transition is invoked the peer is already
+in the [`Connected`](#connected-peer) state.
+The most likely reason is that the router was simultaneously dialing the same
+peer, and has successfully [established a connection](#dialed-transition) with
+it.
+In this case, the transition fails, indicating to the router that is should
+close the accepted connection.
+
+It may also occur that the node is already connected to `MaxConnected` peers,
+which means that all connection slots are full.
+In this case, the peer manager tries to find a connection slot that can be
+[upgraded](#slots-upgrading) to give room for the accepted connection.
+If no suitable connection slot is found, or the hard limit `MaxConnected +
+MaxConnectedUpgrade` of connected peers is reached, the transitions fails.
```

Comment:
Two comments:
1. I find this algorithm very clear. Consider including it in the spec text
2. My initial question was referring to step 4. But we way you word it in your algorithm is now clearer to me (you only check the hard limit once you previously found an "upgradeable" slot, not before)

Reply:
No, we don't. 

But note that once a peer is marked to be evicted, this should not take that long: the router should receive the peer to evict and force the associated threads to stop; when one of the threads quits, the peer is marked as disconnected. It should be worthy know how long it takes, but it should not be long.

In other words, we should not have more than `MaxConnected` connected peers for a long period, as the goal of the peer manager is to attend to this soft limit. But your question is indeed valid, because if we allow this limit to be flexible, and peers should be in this intermediate state for few time, there is no much point in having a hard limit...

Reply:
I added this algorithm to the text, in the DialNext, Dialed, and Accepted transitions.